File: LLD/stategy/blockchain.strategy.cpp
#include<iostream>
using namespace std;



class SignatureAlgo{
    public:
    virtual string sign(string walletAddress, string nonce)=0;
    virtual ~SignatureAlgo(){

    }
};


class Wallet{
    public:
    virtual void setWalletAddress(string walletAddress)=0;
    virtual string getWalletAddress()=0;
    virtual ~Wallet(){
        
    }
};



class Metamask:public Wallet{
    public:
    string walletAddress;

    Metamask(string walletAddress){
        this->setWalletAddress(walletAddress);
    }

    void setWalletAddress(string walletAddress){
        this->walletAddress = walletAddress;
    }

    string getWalletAddress(){
        return this->walletAddress;
    }
};


class Pera:public Wallet{
    public:
    string walletAddress;
    void setWalletAddress(string walletAddress){
        this->walletAddress = walletAddress;
    }

    string getWalletAddress(){
        return this->walletAddress;
    }
};






class EthereumSigner: public SignatureAlgo{
    public:
    string sign(string walletAddress, string nonce){
        return "EthereumSignerSignature"+walletAddress+nonce;
    }
};


class PolygonSigner: public SignatureAlgo{
    public:
    string sign(string walletAddress, string nonce){
        return "PolygonSignerSignature"+walletAddress+nonce;
    }
};


class Blockchain{

    public:
        SignatureAlgo *sa;
        Wallet *w;
        string blockchainName;
    
    ~Blockchain(){
        delete sa;
        delete w;
    }


    void setBlockChain(string blockchainName){
        this->blockchainName = blockchainName;
    }

    string getBlockchain(){
        return this->blockchainName;
    }


    string signSignature(string walletAddress, string nonce){
        return sa->sign(walletAddress,nonce);
    }

    string getWalletAddress(){
        return w->getWalletAddress();
    }

    void setSignatureAlgo(SignatureAlgo *algo){
        this->sa = algo;
    }

    void setWallet(Wallet *wallet){
        this->w = wallet;
    } 
    
};


class Ethereum: public Blockchain{
    public:
    Ethereum(){
        this->blockchainName = "ethereum";
    }
};

class BlockChainFactory{
    public:
    static SignatureAlgo* SignerFactory(string blockchain){
        if (blockchain == "ethereum") {
            return new EthereumSigner();
        }else if(blockchain == "polygon"){
            return new PolygonSigner();
        }
        else{
            throw logic_error("wrong blockchain");
        }
    }
};




int main(){

    string walletAddress = "0xksjdfbkjfbeqd";
    string nonce = "ascqw123e2";

    Blockchain *eth = new Ethereum();
    Wallet *w =  new Metamask(walletAddress);

    eth->setSignatureAlgo(BlockChainFactory::SignerFactory(eth->getBlockchain()));
    eth->setWallet(w);

    string signature= eth->signSignature(walletAddress,nonce);

    cout<<signature<<" "<<walletAddress<<endl;


    eth->setSignatureAlgo(BlockChainFactory::SignerFactory("polygon"));
    string sig2 = eth->signSignature(walletAddress, nonce);

    cout<<signature<<" "<<walletAddress<<endl;



    delete eth;

}

File: LLD/stategy/blockchain.strategy.ts
// ─────────────────────────────────────────────
// STRATEGY PATTERN — Blockchain Signature Verification
// miniOrange use case | Nitin Gupta | Interview Prep
// ─────────────────────────────────────────────


// ── 1. Strategy Interface ──────────────────────────────────────────────────
interface SignatureAlgorithm {
    sign(walletAddress: string, nonce: string): string;
  }
  
  
  // ── 2. Wallet Interface ────────────────────────────────────────────────────
  interface Wallet {
    getWalletAddress(): string;
    setWalletAddress(address: string): void;
  }
  
  
  // ── 3. Concrete Wallets ────────────────────────────────────────────────────
  class Metamask implements Wallet {
    private walletAddress: string;
  
    constructor(walletAddress: string) {
      this.walletAddress = walletAddress;
    }
  
    getWalletAddress(): string {
      return this.walletAddress;
    }
  
    setWalletAddress(address: string): void {
      this.walletAddress = address;
    }
  }
  
  class PeraWallet implements Wallet {
    private walletAddress: string = "";
  
    getWalletAddress(): string {
      return this.walletAddress;
    }
  
    setWalletAddress(address: string): void {
      this.walletAddress = address;
    }
  }
  
  
  // ── 4. Concrete Strategies ─────────────────────────────────────────────────
  class EthereumSigner implements SignatureAlgorithm {
    sign(walletAddress: string, nonce: string): string {
      return `eth_signed:${walletAddress}:${nonce}`;
    }
  }
  
  class PolygonSigner implements SignatureAlgorithm {
    sign(walletAddress: string, nonce: string): string {
      return `polygon_signed:${walletAddress}:${nonce}`;
    }
  }
  
  class SolanaSigner implements SignatureAlgorithm {
    sign(walletAddress: string, nonce: string): string {
      return `solana_signed:${walletAddress}:${nonce}`;
    }
  }
  
  
  // ── 5. Signer Registry (Map — NOT if-else) ─────────────────────────────────
  // OCP: adding a new chain = one new line here, zero changes elsewhere
  const signerRegistry: Record<string, () => SignatureAlgorithm> = {
    ethereum: () => new EthereumSigner(),
    polygon:  () => new PolygonSigner(),
    solana:   () => new SolanaSigner(),
  };
  
  
  // ── 6. SignerFactory ───────────────────────────────────────────────────────
  class SignerFactory {
    static getStrategy(chain: string): SignatureAlgorithm {
      const factory = signerRegistry[chain];
      if (!factory) throw new Error(`Unsupported chain: ${chain}`);
      return factory();
    }
  
    // OCP: register new chains at runtime without touching existing code
    static register(chain: string, factory: () => SignatureAlgorithm): void {
      signerRegistry[chain] = factory;
    }
  }
  
  
  // ── 7. Context Class ───────────────────────────────────────────────────────
  class BlockchainService {
    private strategy: SignatureAlgorithm;
    private wallet: Wallet;
    private blockchainName: string;
  
    constructor(chain: string, wallet: Wallet) {
      this.blockchainName = chain;
      this.wallet = wallet;
      this.strategy = SignerFactory.getStrategy(chain); // DIP: depends on interface
    }
  
    // Core of Strategy Pattern — swap algorithm at runtime
    setStrategy(chain: string): void {
      this.blockchainName = chain;
      this.strategy = SignerFactory.getStrategy(chain);
    }
  
    verifyAccess(nonce: string): string {
      const address = this.wallet.getWalletAddress();
      return this.strategy.sign(address, nonce);
    }
  
    getChain(): string {
      return this.blockchainName;
    }
  }
  
  
  // ── 8. Usage / Main ────────────────────────────────────────────────────────
  const walletAddress = "0xABC123DEF456";
  const nonce = "secure_nonce_xyz";
  
  const wallet = new Metamask(walletAddress);
  const service = new BlockchainService("ethereum", wallet);
  
  // Ethereum signing
  const sig1 = service.verifyAccess(nonce);
  console.log("Ethereum:", sig1);
  // → eth_signed:0xABC123DEF456:secure_nonce_xyz
  
  // Swap to Polygon at runtime — no class changes, no if-else
  service.setStrategy("polygon");
  const sig2 = service.verifyAccess(nonce);
  console.log("Polygon:", sig2);
  // → polygon_signed:0xABC123DEF456:secure_nonce_xyz
  
  // OCP in action: add new chain — zero existing code modified
  SignerFactory.register("avalanche", () => ({
    sign: (addr, n) => `avalanche_signed:${addr}:${n}`
  }));
  service.setStrategy("avalanche");
  const sig3 = service.verifyAccess(nonce);
  console.log("Avalanche:", sig3);
  // → avalanche_signed:0xABC123DEF456:secure_nonce_xyz